### PR TITLE
added apptainer easyblock

### PR DESCRIPTION
**Pull request by mboisson**
I would like your feedback on this

### CODE REVIEW

File: easybuild/easyblocks/generic/apptainer.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+            'source_param': [None, "Source parameter to pass to the build script.", CUSTOM],
+            'source_type': [None, "Type of source used by the build script", CUSTOM],
+            'apptainer_params': ["", "Default parameters for apptainer", CUSTOM],
+        })
+        return extra_vars
+
+    def __init__(self, *args, **kwargs):
+        """Initialize custom class variables."""
+        super(Apptainer, self).__init__(*args, **kwargs)
+        self.containerpath = os.path.join("/cvmfs/containers.computecanada.ca/content/containers", "-".join([self.name.lower(), self.version]))
```
Comment (ccoulombe) [inline easybuild/easyblocks/generic/apptainer.py]: If it is to be used upstream, this should be configurable and not hardcoded. 
I think that in all cases, it should be configurable though.

Comment (mboisson) [reply easybuild/easyblocks/generic/apptainer.py]: It is not to be used upstream. The path is not configurable by design for the build script. That is just the result of where the build script puts the container.

Comment (ccoulombe) [reply easybuild/easyblocks/generic/apptainer.py]: Ok, I'd still think this path should be set elsewhere as a good practice of not setting full path inside code line deep inside a framework.

**Review by ccoulombe — COMMENTED** [review]
Can a sanity check command, for instance `X --version` from a binary from the container be defined in the eb config ?

**Review by ccoulombe — APPROVED** [review]
Otherwise, this lgtm

File: easybuild/easyblocks/generic/apptainer.py (hunk 1 above)
Comment (bartoldeman) [reply easybuild/easyblocks/generic/apptainer.py]: yes I agree with Charles, hardcoding the string so deep just doesn't smell right.
Just a question where you to put it, I'm not sure yet.

File: easybuild/easyblocks/generic/apptainer.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with EasyBuild.  If not, see <http://www.gnu.org/licenses/>.
+##
+"""
+General EasyBuild support for software, using containerized applications
+
+@author: Maxime Boissonneault (Universitle Laval, Calcul Quebec, Digital Research Alliance of Canada)
```
Comment (bartoldeman) [inline easybuild/easyblocks/generic/apptainer.py]: Université :)

**Comment by mboisson** [comment]
> Can a sanity check command, for instance `X --version` from a binary from the container be defined in the eb config ?

no, EasyBuild runs commands, not aliases.

File: easybuild/easyblocks/generic/apptainer.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+        """
+        Don't extend PATH/LIBRARY_PATH/etc.
+        """
+        return ""
+
+    def make_module_step(self, fake=False):
+        """
+        Custom module step for Apptainer: use container path directly
+        """
+        # For module file generation: temporarly set Nix profile
```
Comment (bartoldeman) [inline easybuild/easyblocks/generic/apptainer.py]: Nix?

File: easybuild/easyblocks/generic/apptainer.py (hunk 1 above)
Comment (mboisson) [reply easybuild/easyblocks/generic/apptainer.py]: So, the equivalent for nix was the nix profile path, which was hardcoded in each EasyConfig: 
```
nix_profile = '/cvmfs/soft.computecanada.ca/nix/var/nix/profiles/python-%(version)s'
```

Comment (bartoldeman) [reply easybuild/easyblocks/generic/apptainer.py]: simplest would be to just put
DEFAULT_CONTAINERPATH = /cvmfs/containers.computecanada.ca/content/containers

File: easybuild/easyblocks/generic/apptainer.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+        """No configuration, included in the container recipe"""
+        pass
+
+    def build_step(self):
+        """Compilation done in install step"""
+        pass
+
+    def install_step(self):
+        """Build and install the container."""
+        cmd = "/bin/sudo -iu containeruser build_container_image.sh -t sandbox -v %s -n %s -s %s -i %s" % (self.version, self.name.lower(), self.cfg['source_param'], self.cfg['source_type'])
```
Comment (bartoldeman) [inline easybuild/easyblocks/generic/apptainer.py]: this could be configurable as well

**Review by bartoldeman — CHANGES_REQUESTED** [review]

File: easybuild/easyblocks/generic/apptainer.py (hunk 4 above)
Comment (mboisson) [reply easybuild/easyblocks/generic/apptainer.py]: configurable how ? 

short of having a parameter `run_this_command_to_install` with whatever people want to put in there (which I don't think we should do)

File: easybuild/easyblocks/generic/apptainer.py (hunk 1 above)
Comment (bartoldeman) [reply easybuild/easyblocks/generic/apptainer.py]: yes the easyconfig is where we are supposed to put parameters.

perhaps a better solution is to define a new toolchain and use the MNS to define the install location, but that's something quite different! And not something I'd burn myself 24hrs before leaving on vacation.

Comment (mboisson) [reply easybuild/easyblocks/generic/apptainer.py]: I could make it an argument and then set it through hooks ?

File: easybuild/easyblocks/generic/apptainer.py (hunk 3 above)
Comment (mboisson) [reply easybuild/easyblocks/generic/apptainer.py]: fixed in next commit

File: easybuild/easyblocks/generic/apptainer.py (hunk 2 above)
Comment (mboisson) [reply easybuild/easyblocks/generic/apptainer.py]: fixed in next commit

File: easybuild/easyblocks/generic/apptainer.py (hunk 1 above)
Comment (mboisson) [reply easybuild/easyblocks/generic/apptainer.py]: The install location is not dictated by EasyBuild, it is dictated by the container build script.

File: easybuild/easyblocks/generic/apptainer.py (hunk 4 above)
Comment (mboisson) [reply easybuild/easyblocks/generic/apptainer.py]: I think the way to make this configurable is to code the whole building of the container within the EasyBlock and not use an external script, but that is a much larger project that I'm not willing to undertake unless there is some serious interest from upstream.

Comment (bartoldeman) [reply easybuild/easyblocks/generic/apptainer.py]: You can make this use the `binary.py` easyblock which has
` cmd = ' '.join([self.cfg['preinstallopts'], install_cmd, self.cfg['installopts']]`
then put the info in the easyconfig like this:
```
install_cmd = "/bin/sudo -iu containeruser build_container_image.sh"
installopts = "-v %(version)s -n %(namelower)s -s ... -i... "
```
Now if you prefer not to copy `install_cmd` in all easyconfigs that use this, I'd just put it as a constant at the top of the easyblock for now.

File: easybuild/easyblocks/generic/apptainer.py (hunk 1 above)
Comment (mboisson) [reply easybuild/easyblocks/generic/apptainer.py]: moved it to the easyconfig

File: easybuild/easyblocks/generic/apptainer.py (hunk 4 above)
Comment (mboisson) [reply easybuild/easyblocks/generic/apptainer.py]: ok, converted to the Binary EasyBlock, but with a default command that is not changeable (at least until we get better).

**Pull request merged**